File: benchmarks/haskell/binary_trees.hs
-- Binary trees benchmark for Haskell

import System.Environment
import System.IO
import Control.Concurrent
import Control.Parallel
import Control.Monad
import Data.Time.Clock
import Text.Printf

data Tree = Node !Int !Tree !Tree | Empty

-- Create a binary tree of a given depth
makeTree :: Int -> Tree
makeTree depth
    | depth <= 0 = Node 1 Empty Empty
    | otherwise  = Node 1 (makeTree (depth - 1)) (makeTree (depth - 1))

-- Check a tree by calculating its item sum
checkTree :: Tree -> Int
checkTree Empty = 0
checkTree (Node item left right) = item + checkTree left + checkTree right

-- Process a tree of given depth
iterateTree :: Int -> Int
iterateTree depth = checkTree $ makeTree depth

-- Create and check a bunch of trees
mainProcess :: Int -> Int -> Int -> IO ()
mainProcess minDepth maxDepth stretchDepth = do
    -- Create and check a stretch tree
    let stretchTree = makeTree stretchDepth
    let stretchCheck = checkTree stretchTree
    printf "stretch tree of depth %d\t check: %d\n" stretchDepth stretchCheck

    -- Create and check a long-lived tree
    let longLivedTree = makeTree maxDepth
    
    -- Create and check multiple trees of increasing depth
    let depths = [minDepth, minDepth + 2 .. maxDepth]
    mapM_ (\depth -> do
        let iterations = 2 ^ (maxDepth - depth + minDepth)
        let check = sum [iterateTree depth | _ <- [1..iterations]]
        printf "%d\t trees of depth %d\t check: %d\n" iterations depth check
        ) depths
    
    -- Check the long-lived tree at the end
    let longLivedCheck = checkTree longLivedTree
    printf "long lived tree of depth %d\t check: %d\n" maxDepth longLivedCheck

main :: IO ()
main = do
    -- Define tree depths
    let n = 10  -- Default value
    let minDepth = 4
    let maxDepth = max (minDepth + 2) n
    let stretchDepth = maxDepth + 1
    
    startTime <- getCurrentTime
    
    -- Process the trees
    mainProcess minDepth maxDepth stretchDepth
    
    endTime <- getCurrentTime
    let diff = diffUTCTime endTime startTime
    printf "Time taken: %s ms\n" (show (diff * 1000))

File: benchmarks/swift/binary_trees.swift
// BinaryTrees benchmark for Swift

import Foundation

// Tree node class
class TreeNode {
    var left: TreeNode?
    var right: TreeNode?
    var item: Int
    
    init(item: Int, left: TreeNode? = nil, right: TreeNode? = nil) {
        self.item = item
        self.left = left
        self.right = right
    }
}

// Create a binary tree of specified depth
func makeTree(depth: Int, item: Int = 0) -> TreeNode {
    if depth <= 0 {
        return TreeNode(item: item)
    } else {
        return TreeNode(
            item: item,
            left: makeTree(depth: depth - 1, item: 2 * item - 1),
            right: makeTree(depth: depth - 1, item: 2 * item)
        )
    }
}

// Check a tree by calculating its sum
func checkTree(_ node: TreeNode?) -> Int {
    guard let node = node else { return 0 }
    return node.item + checkTree(node.left) - checkTree(node.right)
}

func main() {
    let minDepth = 4
    let maxDepth = 12
    let stretchDepth = maxDepth + 1
    
    let startTime = Date()
    
    // Create and check stretch tree
    let stretchTree = makeTree(depth: stretchDepth)
    print("stretch tree of depth \(stretchDepth)\t check: \(checkTree(stretchTree))")
    
    // Create long-lived tree
    let longLivedTree = makeTree(depth: maxDepth)
    
    // Process trees of increasing depth
    for depth in stride(from: minDepth, through: maxDepth, by: 2) {
        let iterations = 1 << (maxDepth - depth + minDepth)
        var check = 0
        
        for i in 0..<iterations {
            let a = makeTree(depth: depth, item: i)
            let b = makeTree(depth: depth, item: -i)
            check += checkTree(a) + checkTree(b)
        }
        
        print("\(iterations * 2)\t trees of depth \(depth)\t check: \(check)")
    }
    
    // Check long-lived tree
    print("long lived tree of depth \(maxDepth)\t check: \(checkTree(longLivedTree))")
    
    let elapsedTime = -startTime.timeIntervalSinceNow * 1000
    print("Time taken: \(elapsedTime) ms")
}

main()